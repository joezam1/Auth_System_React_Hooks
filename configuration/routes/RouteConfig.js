const routeConfiguration = Object.freeze({
    homePath:'/',
    authLoginPath:'/auth/login',
    authRegisterPath:'/auth/register',
    authLogoutPath:'/auth/logout',
    privateCustomerDashboardPath:'/private/customer/dashboard',
    privateCustomerOrdersPath:'/private/customer/orders',
    privateCustomerSilverOffersPath:'/private/customer/silverOffers',
    privateCustomerGoldOffersPath:'/private/customer/goldOffers',
    privateCustomerPremiumOffersPath:'/private/customer/premiumOffers',
    privateCustomerPremiumClubMemberPath:'/private/customer/premiumClubMember',
    privateCustomerNewsletterPath:'/private/customer/newsletter',
    privateCustomerManagerOrdersPath:'/private/customer/ManagerOrders',
    privateCustomerContactSuppliersPath:'/private/customer/contact/suppliers',
    storeOrdersPath:'/store/orders',
    storeInvoicesPath:'/store/invoices',
    storeWarehousePath:'/store/warehouse',
    storeAdvertisingPath:'/store/advertising'

});

export default routeConfiguration;