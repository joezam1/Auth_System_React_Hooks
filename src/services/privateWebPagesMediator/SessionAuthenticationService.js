import EnvConfig from '../../../configuration/environment/EnvConfig.js';
import RouteConfig from '../../../configuration/routes/RouteConfig.js';
import ServerConfig from '../../../configuration/server/ServerConfig.js';
import CookieProperty from '../../library/stringLiterals/CookieProperty.js';
import CookieService from '../cookieStorage/CookieService.js';
import LocalStorageService from '../localStorage/LocalStorageService.js';
import inputCommonInspector from '../validators/InputCommonInspector';

import WebWorkerManager from '../../backgroundWorkers/WebWorkerManager.js';
import FetchWorker from '../../backgroundWorkers/FetchWorker';

import FetchWorkerHelper from '../../backgroundWorkers/FetchWorkerHelper.js';
import HttpRequestMethod from '../../library/enumerations/HttpRequestMethod.js';
import WindowLocationProperty from '../../library/stringLiterals/WindowLocationProperty.js';
import Helpers from '../../library/common/Helpers.js';
import BackgroundWorker from '../../library/enumerations/BackgroundWorker.js';



const SessionAuthenticationService = (function(){
    //Test:DONE
    const redirectPrivateWebpagesMediator = function(redirectTo){
        console.log('redirectPrivateWebpagesMediator-redirectTo', redirectTo);
        saveRedirectValueToLocalStorage(redirectTo);
        let cookieName = LocalStorageService.getItemFromLocalStorage(CookieProperty.NAME);
        let sessionToken = CookieService.getCookieFromDataStoreByName(cookieName);
        console.log('redirectPrivateWebpagesMediator-sessionToken', sessionToken);
        if( !inputCommonInspector.stringIsNullOrEmpty(sessionToken)){
            resolveSessionAuthenticationUsingWebWorkers( sessionToken);
        }
        else{

            console.log('redirectPrivateWebpagesMediator-LOGOUT TRIGGERED');
            Helpers.setUrlRedirect( RouteConfig.authLogoutPath );
        }
    }

    return {
        redirectPrivateWebpagesMediator : redirectPrivateWebpagesMediator,
    }
})();

export default SessionAuthenticationService;

//#REGION Private Functions

function saveRedirectValueToLocalStorage( redirectTo ){
    LocalStorageService.removeItemFromLocalStorage(WindowLocationProperty.REDIRECT);
    LocalStorageService.setItemInLocalStorage( WindowLocationProperty.REDIRECT, redirectTo );

    let _redirectValue = LocalStorageService.getItemFromLocalStorage( WindowLocationProperty.REDIRECT );
    if(!inputCommonInspector.inputExist(_redirectValue)){
        console.log('WARNING-RedirectTo-value not SET:', _redirectValue);
        return;
    }
    console.log('RedirectTo-value SET-OK:', _redirectValue);
}

function resolveSessionAuthenticationUsingWebWorkers(currentSessionToken){
    WebWorkerManager.createNewWorker(FetchWorker, authenticateSessionFetchWorkerMessageCallback);
    let apiRequestMessage = getMessageForApiSession(currentSessionToken);
    console.log('resolveSessionAuthenticationUsingWebWorkers-apiRequestMessage', apiRequestMessage);
    WebWorkerManager.sendMessageToWorker(apiRequestMessage);
}


function getMessageForApiSession(cookieValue){
    var sessionUrl = EnvConfig.PROTOCOL +'://' + EnvConfig.TARGET_URL + ServerConfig.apiSessionsSessionTokenGet;
    let requestMethod = HttpRequestMethod[HttpRequestMethod.GET];
    let headersArray = [{name:'x_session_id', value: cookieValue }];
    let payload = {};
    let message = FetchWorkerHelper.getMessageDataForFetchWorker(sessionUrl, requestMethod, headersArray, payload);
    return message;
}


function authenticateSessionFetchWorkerMessageCallback(event){
    console.log('sessionWorkerOnMessageCallback-event', event)
    let sessionInfo = event?.data?.data?.result;

    console.log('sessionWorkerOnMessageCallback-event?.data?', event?.data)
    console.log('sessionWorkerOnMessageCallback-event?.data?.data?', event?.data?.data)
    console.log('sessionWorkerOnMessageCallback-event?.data?.data?.result', event?.data?.data?.result)
    console.log('sessionWorkerOnMessageCallback-sessionInfo', sessionInfo)



    console.log('sessionWorkerOnMessageCallback-sessionInfo', sessionInfo)
    if(inputCommonInspector.objectIsValid(sessionInfo)){
        let apiStoredSessionToken = sessionInfo?.sessionToken?.fieldValue;
        let cookieName = LocalStorageService.getItemFromLocalStorage( CookieProperty.NAME);
        let currentSessionCookie = CookieService.getCookieFromDataStoreByName(cookieName);

        console.log('NEW apiStoredSessionToken', apiStoredSessionToken)
        console.log('CURRENT currentSessionCookie', currentSessionCookie)

        if(apiStoredSessionToken === currentSessionCookie){
            let currentLocationRedirect = LocalStorageService.getItemFromLocalStorage(WindowLocationProperty.REDIRECT);
            console.log('authenticateSessionFetchWorkerMessageCallback-currentLocationRedirect:' , currentLocationRedirect);

            if(inputCommonInspector.stringIsValid(currentLocationRedirect)){

                console.log('LOCATION REDIRECT IS VALID-currentLocationRedirect', currentLocationRedirect);

                Helpers.setUrlRedirect( currentLocationRedirect );
            }
            return;
        }
    }

     //we Verify the CALLBACK RESPONSE is from the backend API
     let fetchWorker = BackgroundWorker[BackgroundWorker.FetchWorker];
     if( inputCommonInspector.inputExist(event?.data?.name)  && event?.data?.name.includes( fetchWorker ) && event?.data?.statusText !== '' ){

        console.log('it is a verified and valid request from the API.');
         //No ACTIVE Session then LOGOUT
        console.log('authenticateSessionFetchWorkerMessageCallback-LOGOUT TRIGGERED');
        Helpers.setUrlRedirect( RouteConfig.authLogoutPath );;
     }


}

//#ENDREGION Private Functions