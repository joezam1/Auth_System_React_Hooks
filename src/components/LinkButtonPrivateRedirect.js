import React , { useEffect , useState} from "react";
import SessionAuthenticationService from "../services/privateWebPagesMediator/SessionAuthenticationService";
import InputCommonInspector from '../services/validators/InputCommonInspector.js';
import InputTypeInspector from '../services/validators/InputTypeInspector.js';
import ComponentAuthorizationService from "../services/privateWebPagesMediator/ComponentAuthorizationService";

//Test: DONE
export default function LinkButtonPrivateRedirect(props){
    const [authorization, setAuthorization] = useState('hide');

    useEffect(()=>{
        let roleIsAuthorized = ComponentAuthorizationService.roleIsAuthorized(LinkButtonPrivateRedirect.name);
        console.log('LinkButtonPrivateRedirect-useEffect-roleIsAuthorized', roleIsAuthorized);
        let routeIsAuthorized = ComponentAuthorizationService.routeIsAuthorized(props.redirectToLocation);
        console.log('LinkButtonPrivateRedirect-useEffect-routeIsAuthorized', routeIsAuthorized);
        if(roleIsAuthorized && routeIsAuthorized){
            setAuthorization('block')
        }
    },[])


    function executePrivateRedirect(){
        let redirect = props.redirectToLocation;
        if(InputTypeInspector.isTypeString(redirect) && InputCommonInspector.stringIsValid( redirect)){
            SessionAuthenticationService.redirectPrivateWebpagesMediator( redirect );
        }
    }

    return(<div className={"button-private-redirect-section " + authorization}>
        <button type="button" onClick={()=>{ executePrivateRedirect();  }} data-testid="button-private-redirect-id"> {props.buttonText} </button>
    </div>)
}