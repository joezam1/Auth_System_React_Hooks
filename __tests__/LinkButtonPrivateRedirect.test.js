import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render, cleanup, fireEvent } from '@testing-library/react';

import LinkButtonPrivateRedirect from '../src/components/LinkButtonPrivateRedirect.js';
import RouteConfig from '../configuration/routes/RouteConfig.js';
import SessionAuthenticationService from '../src/services/privateWebPagesMediator/SessionAuthenticationService'

jest.mock('../src/services/privateWebPagesMediator/SessionAuthenticationService.js');


describe('File: LinkButtonPrivateRedirect.js', function(){
    afterEach(cleanup);
    beforeEach(()=>{
        jest.resetAllMocks();
    });
    afterAll(()=>{jest.resetAllMocks(); });

    describe('Component LinkButtonPrivateRedirect', function(){
        let container;
        beforeEach(()=>{
            container = render( <MemoryRouter> <LinkButtonPrivateRedirect/>  </MemoryRouter>  );
        });

        test('It should take a snapshot', function(){
            expect(container.asFragment(<LinkButtonPrivateRedirect/>)).toMatchSnapshot();
        });



    });


    describe('Component: LinkButtonPrivateRedirect With Props', function(){

        test('It can take a snapshot', function(){
            let container1 = render( <MemoryRouter> <LinkButtonPrivateRedirect redirectToLocation={RouteConfig.privateCustomerDashboardPath} buttonText="test" /> </MemoryRouter>);
            expect(container1.asFragment(<LinkButtonPrivateRedirect redirectToLocation={RouteConfig.privateCustomerDashboardPath} buttonText="test"/>)).toMatchSnapshot();
        });


        test('Button Triggers SessionAuthenticationService', function(){
            //Arrange
            let container2 = render( <MemoryRouter> <LinkButtonPrivateRedirect redirectToLocation={RouteConfig.privateCustomerDashboardPath} buttonText="test" /> </MemoryRouter>);
            const button = container2.getByTestId('button-private-redirect-id');
            SessionAuthenticationService.redirectPrivateWebpagesMediator = jest.fn();
            //Act
            fireEvent.click(button);
            //Assert
            expect( SessionAuthenticationService.redirectPrivateWebpagesMediator).toHaveBeenCalledTimes(1);
        });
    });

    describe('Component: LinkButtonPrivateRedirect Without Props', function(){

        test('It can take a snapshot', function(){
            let container3 = render( <MemoryRouter> <LinkButtonPrivateRedirect /> </MemoryRouter>);
            expect(container3.asFragment(<LinkButtonPrivateRedirect/>)).toMatchSnapshot();
        });

        test('ON CLICK Button DOES NOT Trigger SessionAuthenticationService', function(){
            //Arrange
            let container4 = render( <MemoryRouter> <LinkButtonPrivateRedirect /> </MemoryRouter>);
             const button = container4.getByTestId('button-private-redirect-id');
             SessionAuthenticationService.redirectPrivateWebpagesMediator = jest.fn();
            //Act
            fireEvent.click(button);
            //Assert
            expect( SessionAuthenticationService.redirectPrivateWebpagesMediator).toHaveBeenCalledTimes(0);
        });

    });
});