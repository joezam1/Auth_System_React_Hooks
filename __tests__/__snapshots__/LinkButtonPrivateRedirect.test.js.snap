// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File: LinkButtonPrivateRedirect.js Component LinkButtonPrivateRedirect It should take a snapshot 1`] = `
<DocumentFragment>
   
  <div
    class="button-private-redirect-section"
  >
    <button
      data-testid="button-private-redirect-id"
      type="button"
    >
        
    </button>
  </div>
    
</DocumentFragment>
`;

exports[`File: LinkButtonPrivateRedirect.js Component: LinkButtonPrivateRedirect With Props It can take a snapshot 1`] = `
<DocumentFragment>
   
  <div
    class="button-private-redirect-section"
  >
    <button
      data-testid="button-private-redirect-id"
      type="button"
    >
       test 
    </button>
  </div>
   
</DocumentFragment>
`;

exports[`File: LinkButtonPrivateRedirect.js Component: LinkButtonPrivateRedirect Without Props It can take a snapshot 1`] = `
<DocumentFragment>
   
  <div
    class="button-private-redirect-section"
  >
    <button
      data-testid="button-private-redirect-id"
      type="button"
    >
        
    </button>
  </div>
   
</DocumentFragment>
`;
